Automattic/sensei
# Add theme style support to Learning Mode

Fixes #4760

### Changes proposed in this Pull Request

- [x] Enable block templates independent of theme override
- [x] Add setting to enable theme styles. 
- [x] Skip redirect to `/learn` if theme styles are enabled
- [x] Don't override theme in editor if theme styles are enabled
- [x] Separate content styles not needed when theme styles are on
- [x] Review CSS for top themes and add `!important` everywhere
- [x] Add hooks to load custom CSS
- [x] WordPress 5.7 support
- [x] Add usage tracking

Other:
- Rename a few files and methods where different meanings of 'theme' was getting confusing

### Themes Checked

- [x] Divi
- [x] Astra
- [x] GeneratePress
- [x] Storefront
- [x] Hello Elementor
- [x] Twenty Twenty-One
- [x] OceanWP
- [x] Twenty Twenty
- [x] Twenty Twenty Two

### Testing instructions

- Open a lesson that has Learning Mode and make sure it still loads with the Sensei course theme
- Go to Sensei settings > Course > Learning Mode, check `Enable theme styles`
- Open a lesson that has LM. Should open without `/learn/` URL prefix, and use the active theme of the styles
- Check that things are not too broken in all the themes


### New/Updated Hooks

#### Filters:

* `sensei_course_learning_mode_theme_override_enabled` - Filters if the theme should be overriden for learning mode.
* `sensei_course_has_learning_mode_enabled` - Fixed the filtered value not being respected in all cases.

#### Actions:

* `sensei_course_learning_mode_load_theme` - Fires when learning mode is loaded for a page.
* `sensei_course_learning_mode_load_override_styles` - Fires when the override theme styles are loaded for learning mode.
* `sensei_course_learning_mode_override_theme` - Fires when the theme is override is added for learning mode.


### Deprecated Code

* `Sensei_Course_Theme_Option::has_sensei_theme_enabled` renamed to `Sensei_Course_Theme_Option::has_learning_mode_enabled`
* `Sensei_Course_Theme_Option::should_use_sensei_theme` renamed to `Sensei_Course_Theme_Option::should_use_learning_mode`


### Screenshot / Video

#### Theme styles

<img width="1560" alt="image" src="https://user-images.githubusercontent.com/176949/153087049-b01e7600-4449-4470-81e1-8374e8d88917.png">

#### Settings

<img width="697" alt="image" src="https://user-images.githubusercontent.com/176949/153045782-d18aef70-067d-4a16-aea6-719b211e867e.png">
